servicetalk-transport-netty-internal/src/test/java/io/servicetalk/transport/netty/internal/AbstractSslCloseNotifyAlertHandlingTest.java

```diff
@@ -0,0 +1,127 @@
+import static io.servicetalk.transport.netty.internal.CloseHandler.forPipelinedRequestResponse;
+import static io.servicetalk.transport.netty.internal.FlushStrategies.defaultFlushStrategy;
+import static io.servicetalk.transport.netty.internal.OffloadAllExecutionStrategy.OFFLOAD_ALL_STRATEGY;
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.hasSize;
+import static org.hamcrest.Matchers.is;
+import static org.mockito.Mockito.mock;
+
+abstract class AbstractSslCloseNotifyAlertHandlingTest {
+
+    private static final WireLoggingInitializer WIRE_LOGGING_INITIALIZER =
+            new WireLoggingInitializer("servicetalk-tests-wire-logger", LogLevel.TRACE, () -> true);
+
+    protected static final String BEGIN = "MSG_BEGIN";
+    protected static final String END = "MSG_END";
+
+    @Rule
+    public final Timeout timeout = new ServiceTalkTestTimeout();
+
+    protected final EmbeddedDuplexChannel channel;
+    protected final DefaultNettyConnection<String, String> conn;
+
+    AbstractSslCloseNotifyAlertHandlingTest(boolean isClient) throws Exception {
+        channel = new EmbeddedDuplexChannel(false);
+        final CloseHandler closeHandler = forPipelinedRequestResponse(isClient, channel.config());
+        conn = DefaultNettyConnection.<String, String>initChannel(channel, DEFAULT_ALLOCATOR, immediate(),
+                END::equals, closeHandler, defaultFlushStrategy(), null,
+                WIRE_LOGGING_INITIALIZER.andThen(ch -> ch.pipeline().addLast(new ChannelDuplexHandler() {
+                    @Override
+                    public void channelRead(final ChannelHandlerContext ctx, final Object msg) {
+                        if (BEGIN.equals(msg)) {
+                            closeHandler.protocolPayloadBeginInbound(ctx);
+                        }
+                        ctx.fireChannelRead(msg);
+                        if (END.equals(msg)) {
+                            closeHandler.protocolPayloadEndInbound(ctx);
+                        }
+                    }
+
+                    @Override
+                    public void write(final ChannelHandlerContext ctx, final Object msg, final ChannelPromise promise) {
+                        if (BEGIN.equals(msg)) {
+                            closeHandler.protocolPayloadBeginOutbound(ctx);
+                        }
+                        if (END.equals(msg)) {
+                            closeHandler.protocolPayloadEndOutbound(ctx);
+                            promise.addListener(f -> {
+                                if (f.isSuccess()) {
+                                    closeHandler.protocolPayloadEndOutboundSuccess(ctx);
+                                }
+                            });
+                        }
+                        ctx.write(msg, promise);
+                    }
+                })), OFFLOAD_ALL_STRATEGY, mock(Protocol.class), NoopConnectionObserver.INSTANCE, isClient)
```

Comment:
We previously agreed not to use offloading with embedded channels because it does not have a proper event loop and offloading leads to flaky tests. Actually, the `OFFLOAD_ALL_STRATEGY` in this case is noop, because I pass `immediate()` executor for offloading. I can create a real noop `ExecutionStrategy` to avoid confusion.

Reply:
not suggesting we need to enhance `EmbeddedChannel` to be thread safe. the reason I suggested using different executors was to ensure expectations around dropping data are consistently validated.